Modules/Rpeaks/HilbertTransform.cpp

```diff
@@ -0,0 +1,265 @@
+        double acc = 1. / (2 * fs) * (signal[i + 1] - signal[i]);
+        deriv.push_back(acc);
+    }
+    return deriv;
+}
+
+
+vector<double> HilbertTransform::ComputeHilbertTransform(vector<double> signal, int first) const {
+#define REAL(z, i) ((z)[2*(i)])
+#define IMAG(z, i) ((z)[2*(i)+1])
+
+    double data[2 * SIZE];
+
+    for (int i = 0; i < SIZE; i++) {
+        REAL(data, i) = signal[first + i];
+        IMAG(data, i) = 0.0;
+    }
+
+    // fast fourier transform
+    gsl_fft_complex_radix2_forward(data, 1, SIZE);
+
+    // set DC component to 0
+    REAL(data, 0) = 0.0;
+    IMAG(data, 0) = 0.0;
+
+    for (int i = 0; i < SIZE / 2; i++) {
+        gsl_complex a, b;
+        GSL_REAL(a) = REAL(data, i);
+        GSL_IMAG(a) = IMAG(data, i);
+        GSL_REAL(b) = 0.0;
+        GSL_IMAG(b) = -1.0;
+
+        gsl_complex c = gsl_complex_mul(a, b);
+        REAL(data, i) = GSL_REAL(c);
+        IMAG(data, i) = GSL_IMAG(c);
+    }
+
+
+    for (int i = SIZE / 2; i < SIZE; i++) {
+        gsl_complex a, b;
+        GSL_REAL(a) = REAL(data, i);
+        GSL_IMAG(a) = IMAG(data, i);
+        GSL_REAL(b) = 0.0;
+        GSL_IMAG(b) = 1.0;
+
+        gsl_complex c = gsl_complex_mul(a, b);
+        REAL(data, i) = GSL_REAL(c);
+        IMAG(data, i) = GSL_IMAG(c);
+    }
+
+    // inverse fast fourier transform
+    gsl_fft_complex_radix2_inverse(data, 1, SIZE);
+
+    vector<double> h;
+    for (int i = 0; i < SIZE; i++) {
+        h.push_back(REAL(data, i));
+    }
+
+#undef REAL
+#undef IMAG
+    return h;
+}
+
+
+int HilbertTransform::CalcAverageDistance(vector<int> peaks) const {
+    double distance = 0;
+    for (int i = 1; i < peaks.size(); i++) {
+        distance += (peaks[i] - peaks[i - 1]);
+    }
+    return distance / (peaks.size() - 1);
+}
+
+
+vector<int> HilbertTransform::GetPeaks(vector<double> electrocardiogram_signal) const {
+
+
+    vector<double> signal = Filter(electrocardiogram_signal, 8, 20);
+    signal.erase(signal.begin(), signal.begin() + 9); // offset caused by filters
+    signal = this->Derivative(signal);
+
+    int i = 0;
+    double max_amplitude_old = 0;
+    vector<int> peaks;
+    vector<int> windows;
+    while ((i + SIZE) < signal.size()) {
+
+        windows.push_back(i);
+        vector<double> signal_hilb = ComputeHilbertTransform(signal, i);
+        double rms_value = CalcRMSValue(signal_hilb);
+        auto max_amplitude = max_element(signal_hilb.begin(), signal_hilb.end());
+        double threshold;
+        if (rms_value >= 0.18 * *max_amplitude) {
+            threshold = 0.39 * *max_amplitude;
+            if (*max_amplitude > 2 * max_amplitude_old) {
+                threshold = 0.0039 * max_amplitude_old;
+            }
+        } else { //rms_value < 0.18 * rms_value
+            threshold = 1.6 * rms_value;
+        }
+
+        if (*max_amplitude < threshold) { //peak was not found
+            throw ("Peak was not found");
```

Comment:
Nie jestem pewien czy to dobry pomysł. Musielibyśmy bawić się jeszcze w try catch w view modelu co nam dołoży zbędnej pracy.